pkg/operator/clustermembercontroller/clustermembercontroller.go

```diff
@@ -90,117 +103,322 @@ func (c *ClusterMemberController) reconcileMembers(ctx context.Context, recorder
+		// a discovery error ("error validating peerURLs: ... member count is unequal") because
+		// the unstarted members won't include each other in their ETCD_INITIAL_CLUSTER env.
+		runningNotReady, err := c.isEtcdContainerRunningNotReady(node)
+		if err != nil {
+			errs = append(errs, fmt.Errorf("failed to check if pod is running and not ready: %v", err))
+			continue
+		}
+		if !runningNotReady {
+			continue
+		}
+
+		// We have a valid new node with a running etcd pod to add as a new member
+		return peerURL, kerrors.NewAggregate(errs)
+	}
+	return "", kerrors.NewAggregate(errs)
+}
+
+// ensureEtcdLearnerPromotion checks for all learner members in the cluster membership and tries
+// to promote them if they have the required machine deletion hooks present
+func (c *ClusterMemberController) ensureEtcdLearnerPromotion(ctx context.Context, recorder events.Recorder) error {
+	nodes, err := c.masterNodeLister.List(c.masterNodeSelector)
+	if err != nil {
 		return err
-	} else if machineForHostPendingDeletion {
+	}
+
+	nonVotingMemberNodes, err := c.allNodesMapToVotingMembers(nodes)
+	if err != nil {
+		return fmt.Errorf("failed to map nodes to voting members: %v", err)
+	}
+	// If all nodes already map to voting members then there are no nodes
+	// with learner members awaiting promotion
+	if len(nonVotingMemberNodes) == 0 {
 		return nil
 	}
 
-	recorder.Eventf("FoundPodToScale", "found pod to add to etcd membership: %v", podToAdd.Name)
-	err = c.etcdClient.MemberAdd(ctx, fmt.Sprintf("https://%s", net.JoinHostPort(etcdHost, "2380")))
+	members, err := c.etcdClient.MemberList(ctx)
 	if err != nil {
-		return fmt.Errorf("could not add member :%w", err)
+		return fmt.Errorf("could not get etcd member list: %v", err)
 	}
-	return nil
+
+	// aggregate errors so we don't block on the failure to promote a member
+	var errs []error
+	for _, member := range members {
+		// Check if the member should be promoted
+		promote, err := c.shouldPromote(member)
+		if err != nil {
+			errs = append(errs, fmt.Errorf("failed to decide on promotion for member (%s): %v", member.Name, err))
+			continue
+		}
+		if !promote {
+			continue
+		}
+
+		// Attempt promoting a learner member
+		err = c.etcdClient.MemberPromote(ctx, member)
+		if err != nil {
+			// Note: Cannot use errors.Is(err, etcdserver.ErrLearnerNotReady) as that is always false in this case.
+			// because etcdserver.ErrLearnerNotReady always returns an errors.New() error type
+			// So we compare the error strings here instead
+			if err.Error() == etcdserver.ErrLearnerNotReady.Error() {
+				// Not being ready for promotion is an expected state until the learner catches up
+				klog.V(2).Infof("Not ready for promotion: etcd learner member (%s) is not yet in sync with leader's log ", member.PeerURLs[0])
+				continue
+			}
+			errs = append(errs, fmt.Errorf("failed to promote learner member (%s): %v", member.PeerURLs[0], err))
```

Comment:
we could fire an event, that would allow us to catch potential issues in CI. The event could be throttled every X min

Reply:
We already do have an events for promotion failure or success in the `MemberPromote()` method itself.
```Go
	defer func() {
		if err != nil {
			// Not being ready for promotion can be a common event until the learner's log
			// catches up with the leader, so we don't emit events for failing for that case
			if err.Error() == etcdserver.ErrLearnerNotReady.Error() {
				return
			}
			g.eventRecorder.Warningf("MemberPromote", "failed to promote learner member %s: %v", member.PeerURLs[0], err)
		} else {
			g.eventRecorder.Eventf("MemberPromote", "successfully promoted learner member %v", member.PeerURLs[0])
		}
	}()
``` 
https://github.com/openshift/cluster-etcd-operator/pull/758/files#diff-79a9561715e7c906edbb7c786fdbab2cc3f0e6cd83447a8a4aaa5c3c87b0aa2dR256

Or did you mean something else?